src/vmm/src/guest_config/templates/mod.rs

```diff
@@ -1,9 +1,88 @@
+#[cfg(target_arch = "x86_64")]
+pub type CpuTemplate = x86_64::CpuTemplate;
+/// Type alias agnostic of CPU architecture for custom CPU templates.
+#[cfg(target_arch = "aarch64")]
+pub type CpuTemplate = aarch64::CpuTemplate;
+
+/// Type alias agnostic of CPU architecture for CPU configuration.
+#[cfg(target_arch = "x86_64")]
+pub type CpuConfiguration = crate::guest_config::x86_64::CpuConfiguration;
+/// Type alias agnostic of CPU architecture for CPU configuration.
+#[cfg(target_arch = "aarch64")]
+pub type CpuConfiguration = crate::guest_config::aarch64::CpuConfiguration;
+
+/// Configuration errors for with variants specific for x86.
+#[cfg(target_arch = "x86_64")]
+pub type GuestConfigError = x86_64::Error;
+/// Configuration errors for with variants specific for aarch64.
+#[cfg(target_arch = "aarch64")]
+pub type GuestConfigError = aarch64::Error;
+
 // TODO: Refactor code to merge deserialize_* functions for modules x86_64 and aarch64
 /// Templates module to contain sub-modules for aarch64 and x86_64 templates
 
+/// Function that proxies the function call to the target architecture.
+pub fn create_guest_cpu_config(
+    template: &CpuTemplate,
+    host_config: &CpuConfiguration,
+) -> Result<CpuConfiguration, GuestConfigError> {
+    #[cfg(target_arch = "x86_64")]
+    return x86_64::create_guest_cpu_config(template, host_config);
+    #[cfg(target_arch = "aarch64")]
+    return aarch64::create_guest_cpu_config(template, host_config);
+}
+
+/// CPU configuration wrapper that wraps all
+/// possible configuration options
+/// Default - Static(CpuFeaturesTemplate::None)
+/// Hard-coded - Static(CpuFeaturesTemplate::<any non-none variant>)
+/// User-defined - Custom(<provided by user>)
+#[derive(Debug, PartialEq, Eq)]
+pub enum CpuConfigurationTypes {
+    /// Enum for all hard-coded (static) templates
+    Static(CpuFeaturesTemplate),
+    /// CPU configuration to be created by applying
+    /// a template (modifiers) on top of the host's (and KVM's)
+    /// supported vCPU configuration.
+    Custom(CpuConfiguration),
+}
+
+impl Default for CpuConfigurationTypes {
+    /// Returns a "none" type for CPU configuration.
+    /// Firecracker and KVM defaults will be used to configure vCPUs.
+    fn default() -> Self {
+        CpuConfigurationTypes::Static(CpuFeaturesTemplate::None)
+    }
+}
+
+impl CpuConfigurationTypes {
+    /// Will work out what type of CPU configuration is based on
+    /// configuration provided by user.
+    pub fn from(
+        static_config: Option<CpuFeaturesTemplate>,
+        custom_config: Option<CpuConfiguration>,
+    ) -> Result<Self, GuestConfigError> {
+        if static_config.is_some()
+            && static_config.unwrap() != CpuFeaturesTemplate::None
+            && custom_config.is_some()
+        {
+            return Err(GuestConfigError::ConflictingCpuTemplateConfigured);
+        } else if let Some(static_template) = static_config {
```

Comment:
It looks like we always hit this branch, because `from` is always called with `static_config` as Some (builder.rs:491).

Reply:
I think you are correct, at least given that the static template can be populated optionally. I opted for that route as the caller of the from() function shouldn't always need to care about the static template if a custom template is provided.